Add TagsViewModel to restructure UI and orchestration logic in TagsActivity

dandominguez04 opened the pull request:
Fixes #1371

### Fix
This PR restructure the code of `TagsActivity` to use ViewModels (`TagsViewModel`)  in order to decouple data management logic from the view. 

### Test

#### Automatic Tests
1. Run `TagsActivityTest` inside `androidTest`. It requires to run an on emulator or physical device. 
2. Run `TagsViewModelTest` inside `test`

#### Manual Tests
1. Open Simplenote app
2. Tab on left menu
3. Tab on EDIT tags
4. Try deleting tags without and with notes. It should update the list
5. Try tab on tags to open a dialog to update the tag
6. Try tab on + button to add new tags
7. Try searching tags

### Review
This PR will be reviewed by @ParaskP7 .

### Release
These changes do not require release notes.

dandominguez04 committed:
Added TagItem which going to be used in list adapter for TagsActivity.

dandominguez04 committed:
Added methods allTags and searchTags to TagsRepository that return list of tags.

dandominguez04 committed:
Added base TagsViewModel.

dandominguez04 committed:
Added TagsViewModel to TagsActivity and ViewModelFactory.

dandominguez04 committed:
Implement addTag event in TagsViewModel and TagsActivity.

dandominguez04 committed:
Implement long tab on add tag button in TagsViewModel and TagsActivity.

dandominguez04 committed:
Implement finish activity in TagsViewModel and TagsActivity.

dandominguez04 committed:
Added TagItemAdapter that uncouple Simperium from the list of tags.

dandominguez04 committed:
Use TagItemAdapter in TagsActivity to handle the list of tags.

dandominguez04 committed:
Handle delete and edit events at the row level (tagItem).

dandominguez04 committed:
Implement logic to delete tags.

dandominguez04 committed:
Delete TagsAdapter and async task to delete tags from notes (RemoveTagFromNotesTask).

dandominguez04 committed:
Change private visibility for localAllTags.

dandominguez04 committed:
Handle tags updates in SimperiumTagsRepository

dandominguez04 committed:
Handle update in onActivityResult.

dandominguez04 committed:
Handle search updates to scroll to top of the list.

dandominguez04 committed:
Refactor TagsActivity to improve readability.

dandominguez04 committed:
Remove warnings in code.

dandominguez04 committed:
Remove tags from notes first and then delete the tag.

dandominguez04 committed:
Add comments for pause() in TagsViewModel.

dandominguez04 committed:
Setup TagsViewModelTest.

dandominguez04 committed:
Added tests for start and simple event updates.

dandominguez04 committed:
Added tests for closeSearch in TagsViewModelTest.

dandominguez04 committed:
Added tests for TagsViewModelTest.

dandominguez04 committed:
Avoid flaky tests generated by bugs in runBlockingTest and mockito stubs for suspend functions.

ParaskP7 commented on Simplenote/src/main/java/com/automattic/simplenote/viewmodels/TagsViewModel.kt:
```diff
@@ -0,0 +1,120 @@
+
+    fun start() {
+        viewModelScope.launch {
+            val tagItems = withContext(Dispatchers.IO) { tagsRepository.allTags() }
```
Suggestion (💡): This is quite the debatable opinion, so please take it with a grain of salt, I just want to listen your opinion as well. So, here I go. If it was me I would have moved all the context switching, aka `withContext(...)`, downstream and in this case into the repository.

This kind of pattern helps the caller of the API, in this case the caller of the repository to not have to deal with the internals of where is this piece of logic need to run. Instead the API designer should make sure that this is run with the most appropriate dispatcher, in this case the IO, thus making this case move from the main thread and into a background one.

By doing it this way you make it easier to the `ViewModel`, which will now not have to worry about any context switching. This way, the tests also benefit from not having to worry about it, thus making them more deterministic. For example, the bad practice of `Thread.sleep(...)` in unit tests, in this case, can be totally avoided. Obviously, sometimes you need to include, by design, context switching in your `ViewModel` and then you indeed need to write your tests in such a way to switch the main dispatchers with test dispatcher and avoid `Thread.sleep(...)` as well, but in this case, I think it is even easier because this is not even needed if those are moved to the `Repository` level. Wdyt?

Also, this for example, will remove the need to include `withContext(Dispatchers.IO)` every time the `tagsRepository.allTags()` is called, thus removing some duplication. The author of the `ViewModel` actually gets freed up, cognitively, into having to think about all that.

Let me know your thoughts and let's discuss. 🙏

dandominguez04 replied:
I completely agree with this point. It makes the ViewModel much cleaner and moves the responsibility of context switching where it should be. It also fixed our tests.

Fixed in 87b060c762f390a5421c6821e8c7dc96751c2621.

ParaskP7 replied:
Great, thanks for making this change @dandominguez04 ! 🚀

ParaskP7 commented on Simplenote/src/main/java/com/automattic/simplenote/viewmodels/TagsViewModel.kt:
```diff
@@ -0,0 +1,120 @@
+        }
+
+        jobTagsFlow = viewModelScope.launch {
+            tagsRepository.tagsChanged().collect {
```
1) Warning (⚠️): Consider adding tests for this `collect { ... }` logic.
2) Question (❓): Wouldn't the flow need to run on another context as well?

dandominguez04 replied:
> Wouldn't the flow need to run on another context as well?

Yes! They run inside a bucket, so it is safe to run on the `IO` dispatcher. Fixed in 7f19bb404e880a70dff2e6eb636efdbd88b120e6.

dandominguez04 replied:
> Consider adding tests for this collect { ... } logic.

The test added uses a `MutableSharedFlow` as a return value of `tagsChanged` to be able to manipulate the flow whenever we want to trigger the `collect` method. When we emit values to the flow, this triggers the collect method, which calls `allTags()`. Fixed in b05c4fac4952bf65ef17839ab177df2ddc0c316a.

ParaskP7 replied:
👋 @dandominguez04 !

> Yes! They run inside a bucket, so it is safe to run on the IO dispatcher. Fixed in 7f19bb4.

What about using the `callbackFlow { ... }.flowOn(...)` instead?

> Fixed in b05c4fa.

Thanks for the fix! 🙏

dandominguez04 replied:
Yes, I agree `flowOn(...)` is the right method. Fixed in 41920bd4502cab737ff7381edbe5982b5bee1fc3.

ParaskP7 replied:
👋  @dandominguez04 !

> Yes, I agree flowOn(...) is the right method. Fixed in 41920bd.

Great, thanks for updating that!

ParaskP7 commented on Simplenote/src/main/java/com/automattic/simplenote/viewmodels/TagsViewModel.kt:
```diff
@@ -0,0 +1,120 @@
+        }
+    }
+
+    fun pause() {
```
1) Warning (⚠️): Consider adding tests for this function.
2) Question (❓): Do you think this comment is necessary?

dandominguez04 replied:
> Consider adding tests for this function.

Use a similar test to the test for `collect { ... }` logic with the difference that when we call `pause`, any changes to the tags is not listened, therefore the `UiState` is not updated. Fixed in 79c13d4b073b2f1b200967ea0ee9e67b183ffe63.

dandominguez04 replied:
> Do you think this comment is necessary?

Since `callbackFlow` is a new library and it is not explicit and clear that `cancel()` eventually calls to `awaitClose()`, I think the comment is useful for any reader which has not used `callbackFlow` before.

ParaskP7 replied:
👋 @dandominguez04 !

> Fixed in 79c13d4.

Thanks for the fix! 🙏 

> Since callbackFlow is a new library and it is not explicit and clear that cancel() eventually calls to awaitClose(), I think the comment is useful for any reader which has not used callbackFlow before.

This makes sense, let's keep it then! 👍 Maybe, one last suggestion then would be to move this comment(s) to a Javadoc on top of the `pause()` function itself. But, it is totally up to you at this point.

ParaskP7 commented on Simplenote/src/main/java/com/automattic/simplenote/viewmodels/TagsViewModel.kt:
```diff
@@ -0,0 +1,120 @@
+        if (tagItem.noteCount > 0) {
+            _event.postValue(TagsEvent.DeleteTagEvent(tagItem))
+        } else {
+            deleteTag(tagItem)
```
Warning (⚠️): Consider adding tests for this `deleteTag(...)` branch logic.

dandominguez04 replied:
The test `clickDeleteTagOfTagWithNotesShouldDeleteTagDirectly` hits this branch.

ParaskP7 replied:
👋 @dandominguez04 !

This is interesting, it's been some time now and I don't remember exactly, but can you please confirm that by running `Run test with Coverage`? I am pretty sure that when I did that I was seeing this branch, either the `if` or the `else` not being hit. 🤔

dandominguez04 replied:
You are correct. I was using the wrong tag in the test. Fixed in 4246fd70f037c3011491820632227f84f4dd806b.

ParaskP7 replied:
Thanks for the fix! 🙏

ParaskP7 commented on Simplenote/src/main/java/com/automattic/simplenote/viewmodels/TagsViewModel.kt:
```diff
@@ -0,0 +1,120 @@
+    }
+
+    fun deleteTag(tagItem: TagItem) {
+        viewModelScope.launch(Dispatchers.IO) {
```
Question (❓): Above, you choose to use `withContext(Dispatchers.IO) { ... }` instead of `viewModelScope.launch(Dispatchers.IO) { ... }`, but here you are doing it this way, why did you choose one over the other?

dandominguez04 replied:
If we follow the pattern you described above, then the responsibility of changing the context to IO should be of the callee, not the caller. I will change the `launch` call to use the default dispatcher and change the context in the repo as we did with the `allTags` and `searchTags` methods. 

Fixed in cf6f8ce802019fbe99b0d1a09bc0729e3b1f4b68.

ParaskP7 replied:
Thanks for the change @dandominguez04 ! 🚀

ParaskP7 commented on Simplenote/src/main/java/com/automattic/simplenote/TagsActivity.java:
```diff
@@ -82,48 +119,95 @@ public void onCreate(Bundle savedInstanceState) {
+                    boolean isSearching = uiState.getSearchQuery() != null;
+                    checkEmptyList(isSearching);
+            }
+        }));
```
Minor (🔍): Consider reformatting this to:
```
        viewModel.getUiState().observe(this, uiState ->
                tagItemAdapter.submitList(uiState.getTagItems(), () -> {
                    if (uiState.getSearchUpdate()) {
                        mTagsList.scrollToPosition(0);
                        boolean isSearching = uiState.getSearchQuery() != null;
                        checkEmptyList(isSearching);
                    }
                })
        );
```

dandominguez04 replied:
Fixed in 165d851d1a7900f7406fa715f7bbd07a58967aea.

ParaskP7 commented on Simplenote/src/test/java/com/automattic/simplenote/viewmodels/TagsViewModelTest.kt:
```diff
@@ -0,0 +1,232 @@
+    }
+
+    @Before
+    fun setupTags() {
```
1) Minor (🔍): Why using two `@Before` annotation instead having one entry point to it and then using inner calls to split setup into separate functions? To be honest, I didn't know that you can do something like that. 😅 Is that even deterministic, which one will run first, does that depend on the order of the defined annotated functions?
2) Question (❓): Why did you choose to setup the `tagItems` every time a test runs? For example, if it was me I would have move that to field declaration and made this list immutable as well, like:
```
    private val tagItems = listOf(
        TagItem(Tag("tag1"), 0),
        TagItem(Tag("tag2"), 2),
        TagItem(Tag("tag3"), 5),
        TagItem(Tag("tag4"), 10),
        TagItem(Tag("tag5"), 0),
    )
```
I understand that you need that to be mutable, just in this one test, and in that case I would just create another mutable inner to this single test variable that would contain all these `tagItems` values, but open it to mutability. Wdyt?

dandominguez04 replied:
> Why did you choose to setup the tagItems every time a test runs?

There is not a particular reason. I fixed it in ed892e53be8b775c162bc617388d563460b1cc03.

ParaskP7 commented on Simplenote/src/main/java/com/automattic/simplenote/repositories/SimperiumTagsRepository.kt:
```diff
@@ -48,4 +55,71 @@ class SimperiumTagsRepository(
+    }
+
+    @ExperimentalCoroutinesApi
+    override suspend fun tagsChanged(): Flow<Boolean> = callbackFlow {
```
Praise (❤️): This is a super interesting solution, using the `callbackFlow { ... }` trick like that, kudos. The only thing that would made me feel a bit uncomfortable is the fact that this is still an experimental API, but other than that, nice job.

ParaskP7 reviewed (changes requested):
👋 @dandominguez04  !

I have reviewed and tested this PR as per the instructions, everything works as expected, really nice work here as well, it is very impressive to see the diff here, the before and after, so much complexity has been reduced! 🌟 🌟 🌟 

I have left some warning (⚠️) comments, which are mainly test related, then some minor (🔍) ones as well, few suggestions (💡) and a couple of questions (❓) that I want to discuss with you. In addition to those, I have the below I would like to briefly discuss:
- You did a good job splitting the commits like you did. However, please try to avoid combining migration commits with pure refactor commits. It becomes harder to focus on the migration part when an extra refactor commit get squashed into it. Things can split away this way, don't you think?
- Although you did a good job splitting the commits, I would have loved if that was done feature-by-feature, followed by tests, or even better proceeded by tests. Otherwise it is very hard to follow the changes and I found myself trying to clue commits together in order to make sure I understood the full migration. Adding tests at the end do provide value, but adding tests as you progress with the migration makes it easier to follow, at least from the reviewer's point of view. Wdyt?
- Both the above made it really hard for me to follow the adapter migration. You did an excellent job with this migration, but the committing strategy didn't help me figure out how that was done. I was just presented with the result rather than the path to the migration. Then, I had to work it out myself, to make sure that nothing is missing and I still end up feeling that I might missed bits here and there, even after I was done reviewing this PR. Wdyt?
- Why did you choose the `org.mockito.kotlin` library? I am asking that because in all our other repos, or at least the `WPAndroid` one, we are using the `com.nhaarman.mockitokotlin2` instead. I would prefer to keep this consistent, but I am definitely not opposed to using a different library if that is for the best. For example, I would have personally used `io.mockk` for that matter, but I would have first explain the reasoning behind that, else I would have gone with the flow, using `com.nhaarman.mockitokotlin2` as a starting point. Wdyt?

dmsnell commented on Simplenote/src/main/java/com/automattic/simplenote/repositories/SimperiumTagsRepository.kt:
```diff
@@ -48,4 +55,71 @@ class SimperiumTagsRepository(
+        while (cursor.moveToNext()) {
+            val note = cursor.getObject()
+            val tags = note.tags
+            tags.remove(tag.name)
```
we need to keep in mind the distinction here between a tag's name and its `tagHash`

tags are stored in a note's model with the name, of which there might be multiple "lexical variations" that collapse to the same reference, or hash. for example, tags with different casings are identical as are tags whose only difference is their Unicode normal forms.

it would be nice, especially since we're jumping in between notes and tags here, coupling the two, to add a new method to `Note`, which would remove a tag and maintain all the appropriate logic.

```kt
while (cursor.moveToNext()) {
    var note = cursor.getObject()
    note.removeTag(tag.name)
    note.save()
}
```

```kt
// Note
fun removeTag(tag: Tag) {
    tags = tags.filter { name -> ! tag.sameAs(name) }
}
```

dandominguez04 replied:
Fixed in 0429e79402759be2ba4fd93d060e03173f8126bf. I noticed that when the tag is added, it adds the canonical form. For example, if `tag3` exists and I type `TAG3`, the app changes to `tag3`.

dmsnell replied:
> it adds the canonical form. For example, if tag3 exists and I type TAG3, the app changes to tag3.

we've still been settling these decisions. the goal is to be able to prompt someone that what they are adding as a new tag already exists, and then to give them a choice to rename the existing tag or adopt the existing name.

we did decide that in the individual note view, when listing tags, that those should eventually all shows the canonical name as provided in the `Tag` entity itself, regardless of how it's written in the `Note`'s `tags` property.

dandominguez04 replied:
There is also the case of two different users adding lexical variations of the same tag before synchronizing. For example, user 1 adds `TAG3` and user 2 adds `tag3`. The note will keep the two tags I think.

dmsnell commented on Simplenote/src/main/java/com/automattic/simplenote/repositories/SimperiumTagsRepository.kt:
```diff
@@ -48,4 +55,71 @@ class SimperiumTagsRepository(
+
+    override suspend fun searchTags(query: String): List<TagItem> {
+        val tags = Tag.all(tagsBucket)
+                .where(Tag.NAME_PROPERTY, Query.ComparisonType.LIKE, "%$query%")
```
note that here too we will want to compare not the tag name but the tag hash. the `name` for the tag in the Tag bucket might be a different representation than the name given in the query even though they are the same tag. if we don't do this then we'll miss tags in our searches that should appear.

dandominguez04 replied:
@dmsnell do you have some examples with I can work and test. I think I understand the problem but I would like some samples you have had problems with before so I can add tests to be sure we are covering those cases.

I think something like `Tag2` and `tag2` works now because the `like` query in Simperium is case insensitive. However, I think the cases you mentioned related to Unicode normal forms might not pass.

dmsnell replied:
We'll see if I can do this without GitHub mangling the text. (Edit: nope)

 - `Bu\u{0308}cher` and `B\u{00fc}cher` "Bücher" are the same tag
 - `q\u{0307}\u{0323}` and `q\u{0323}\u{0307}` "q̣̇" are the same tag

If these two cases are handled properly (even if it works with the query itself) then it's probably all fine. It _can_ be difficult to get the right test inputs since text entry systems do like to perform their own normalization (that is, even if copying and pasting, sometimes when you paste two different things they turn into the same thing, and it's hard to tell since they are both visually identical).

dandominguez04 replied:
Hi @dmsnell ,

I have tried both cases:

> Bu\u{0308}cher and B\u{00fc}cher "Bücher" are the same tag
![imagen](https://user-images.githubusercontent.com/81183210/123501107-33148f80-d608-11eb-8888-4612b840f99f.png)

 &nbsp;
 &nbsp; 
  

> q\u{0307}\u{0323} and q\u{0323}\u{0307} "q̣̇" are the same tag
![imagen](https://user-images.githubusercontent.com/81183210/123501041-ea5cd680-d607-11eb-9d2f-ce81af91cc7a.png)

It seems that both cases worked. However,

>  It can be difficult to get the right test inputs since text entry systems do like to perform their own normalization (that is, even if copying and pasting, sometimes when you paste two different things they turn into the same thing

Yes, this made me hesitate a little bit about the results, but let me add an issue to add tests for the `SimperiumTagsRepository` and allow me to investigate more how to format the inputs so we can test this properly (see issue #1403).

dmsnell replied:
Good checking! If it's too hard _and_ especially if this is already broken, let's not worry too much about it. Otherwise, in a test (or for a quick check by modifying the code) we should be able to manually add a new tag with a hard-coded name, as we can do that in Java strings as well.

```java
TagUtils.createTagIfMissing(tagBucket, "q\u0307\u0323");
TagUtils.createTagIfMissing(tagBucket, "q\u0323\u0307");
```

and then we can perform a tag search and inspect the tag bucket/database to see if it created two entities, or maybe not 🤔 it's so confusing sometimes to intentionally inject bad data.

dmsnell commented on Simplenote/src/test/java/com/automattic/simplenote/viewmodels/TagsViewModelTest.kt:
```diff
@@ -0,0 +1,232 @@
+            onBlocking { tagsChanged() }.doReturn(flow{ emit(true) })
+        }
+
+        val updatedTags = tagItems.filter { tagItem -> tagItem.tag.name != "tag2" }
```
this works in the tests because we're using lower-cased ASCII/latin1 tag names, but we should rely on the tag hashing mechanism to avoid hiding regressions in the tag comparison logic.

we should be able to click on a tag shown as `Tag2` and find that the `tag2` tag is also gone. this happens regularly with Simplenote due to how notes reference tags by name and not by reference.

dandominguez04 replied:
As I mentioned above, this behavior should be tested at the repository level. At the ViewModel level, the repo is mocked. The solution for #1403 should have the tests for edge cases on tag names.

dmsnell commented on Simplenote/src/test/java/com/automattic/simplenote/viewmodels/TagsViewModelTest.kt:
```diff
@@ -0,0 +1,232 @@
+        }
+        viewModel.updateOnResult()
+
+        // Flay tests regarding runBlockingTest. I will include this sleep here for now
```
typo: Flay -> Flaky?

dandominguez04 replied:
Yes, typo. But we already removed the sleep call since we fixed the tests in 87b060c762f390a5421c6821e8c7dc96751c2621.

dandominguez04 committed:
Moving withContext to SimperiumTagsRepository methods.

Move all context switching in coroutines to the repository to
change the responsability of switching context to the callee
instead of the caller. This help us fixed some tests as well.

dandominguez04 committed:
Move context switching to SimperiumTagsRepository in deleteTag.

Add withContext(Dispatchers.IO) to deleteTag in
SimperiumTagsRepository and use the default dispatcher where
deleteTag is called.

dandominguez04 committed:
Reformat getUiState() observer in TagsActivity.

dmsnell commented on Simplenote/src/main/java/com/automattic/simplenote/TagsActivity.java:
```diff
@@ -82,48 +119,96 @@ public void onCreate(Bundle savedInstanceState) {
         mEmptyViewImage = emptyView.findViewById(R.id.image);
         mEmptyViewText = emptyView.findViewById(R.id.text);
-        checkEmptyList();
+        checkEmptyList(false);
```
_don't_ check empty list?

dandominguez04 replied:
Check empty list while is not searching. Do you think is confusing?

dmsnell replied:
> Do you think is confusing?

is it not confusing to you? `false` does not convey to me the part about `while is not searching`
I know it's common to use these flags, but I find _boolean blindness_ to be a common confusing style.
it's not like `checkEmptyList` itself is that clear, but is there a way we could remove the anxiety
introduced by an unexplained boolean?

dandominguez04 replied:
Yes, I think we can create a specific method for each behavior and let the caller decide which method to call. Allow me to introduce the change.

dmsnell replied:
it looks like we're not even "checking" anything here, more like updating a label.

doodling…

```java
// as separate methods
public void labelNoTags() {
    setEmptyListImage(…);
    setEmptyListMessage(…);
}

public void labelNoTagSearchResults() {
    if (DisplayUtils…) { … } else { … }

    setEmptyListMessage(…);
}

// as a named value with inherent meaning

public enum TagListContext {
    AllTags,
    SearchResults
}

public void updateEmptyListLabel(TagListContext context) {
    switch (context) {
        case AllTags:
            setEmptyListImage(…);
            setEmptyListMessage(…);
            break;

        case SearchResults:
             …
    }
}
```

of course if we use the `enum` we wouldn't have to switch (no pun intended) from an `if` to a `switch`. we could leave everything else the same. thoughts?

```java
checkEmptyList(AllTags); // vs. checkEmptyList(false);
checkEmptyList(SearchResults); // vs. checkEmptyList(); or checkEmptyList(true);
```

dandominguez04 replied:
I implemented them as separate methods. Fixed in e15ef97222b8dfc62bfe4f03b2dcf4b615ab63ec.

dmsnell commented on Simplenote/src/main/java/com/automattic/simplenote/TagsActivity.java:
```diff
@@ -82,48 +119,96 @@ public void onCreate(Bundle savedInstanceState) {
+            mButtonAdd.performHapticFeedback(HapticFeedbackConstants.LONG_PRESS);
+        }
+
+        Toast.makeText(TagsActivity.this, getString(R.string.add_tag), Toast.LENGTH_SHORT).show();
```
in the past I found it helpful to add a class method for toasting. the toasts were all similar enough that they mostly just required passing the string. possibly the length.

```java
private void showLongAddToast() {
    if (mButtonAdd…) {
        …
    }

    toast(R.string.add_tag);
}

private void toast(@StringRes int resId) {
    toast(resId, Toast.LENGTH_SHORT);
}

private void toast(@StringReg int resId, @Duration int duration) {
    Toast.makeText(TagsActivity.this, resId, duration).show();
}
```

besides removing duplicated code, I felt like this baked it down to the essence of what we're doing and relegated the details to the implementation

dandominguez04 replied:
This is a good idea. I have added just one method since both toasts have the same duration, but these are functions that can be added as utility functions for all the project. Fixed in fc1bf536bd26c1eee274a2c11c2290bfda37568a.

```kotlin
 private void toast(@StringRes int resId) {
        Toast.makeText(TagsActivity.this, getString(resId), Toast.LENGTH_SHORT).show();
 }
```

dmsnell commented on Simplenote/src/main/java/com/automattic/simplenote/repositories/SimperiumTagsRepository.kt:
```diff
@@ -48,4 +57,71 @@ class SimperiumTagsRepository(
+            val tag = cursor.`object`
+            val noteCount: Int = notesBucket
+                    .query()
+                    .where(Note.TAGS_PROPERTY, Query.ComparisonType.EQUAL_TO, tag.name)
```
unfortunately this is like the other places and we probably can't use the database query directly. if we don't compare tag hashes we will get the wrong counts. this has been an annoying bug in the app for a long time, so it's not like it's new, but also, we should avoid introducing new regressions.

dandominguez04 replied:
@dmsnell thanks for the feedback. It is clear that there are some bugs regarding tag names and how we compare them and search them. All these queries were passed directly (copied) from the original activity to keep the same behavior while refactoring. Do you think we should make all these changes in this PR or make separate issues to introduce the changes?

dmsnell replied:
> All these queries were passed directly (copied) from the original activity to keep the same behavior while refactoring

Ugh. Bummer - sorry you have to deal with this!

> Do you think we should make all these changes in this PR or make separate issues to introduce the changes?

Like I said, this app has a lot of issues with tags. A separate PR is fine. Would it be much work to propose that change _before_ this refactor? It doesn't matter too much, but I prefer seeing that we fix something before moving it around to another place, leaving its trail of buggy code in the project history.

dandominguez04 replied:
> Ugh. Bummer - sorry you have to deal with this! 

No problem 😅️. Let's fix them all!

> It doesn't matter too much, but I prefer seeing that we fix something before moving it around to another place, leaving its trail of buggy code in the project history.

We could do it before on a separate PR, although I am inclining on doing the changes in this PR since we have the code encapsulated in the SimperiumTagsRepository and the changes might not be that big (few lines of code). 

Let me see if the solution is simple enough to fix it in the repository.

dandominguez04 committed:
Improve readability of methods that set empty list label

The checkEmptyList was refactored into two separate methods to
handle setting the empty list label when there are not tags
in the search results and when the user has not created tags

Commit:
Removed second @Before method from TagsViewModelTest to avoid confusion

Made tagItems in TagsViewModelTest immutable and delete second @Before
method from the test. This helps to avoid setting up the same variable
for each test when we just need it mutable in one test.

Commit:
Update clickDeleteTagOfTagWithoutNotesShouldDeleteTagDirectly to test deleteTag branch

The test clickDeleteTagOfTagWithoutNotesShouldDeleteTagDirectly was not reaching the
branch that delete the tag directly because we were using "tag2" which has 2 notes
associated. We change the test to use "tag1" which has zero notes associated

dandominguez04 committed:
Changed the context in which tagsChanged is executed

The method tagsChanged inside SimperiumTagsRepository now runs on
Dispatcher.IO following the patter we implemented in other methods
in the repository such as allTags and searchTags

dandominguez04 committed:
Added test to trigger tagsChanged flow in TagsViewModel

Added the test whenTagsChangedWithANewTagUiStateShouldUpdate which
triggers the flow tagsChanged. This results in a call to allTags
from the TagsViewModel to the repository.

dandominguez04 committed:
Added test to verify when TagsViewModel execute pause

When pause is called in TagsViewModel, any changes to tags do not
trigger the collect method which calls again allTags. This happens
because the viewmodel is not listening to tag changes anymore.

dandominguez04 committed:
Changed withContext to flowOn on callbackFlow

It it better to use flowOn on Kotlin's flows to define the dispatcher in
which the coroutine is going to be run.

ParaskP7 reviewed (approved):
👋 @dandominguez04 !

I have looked at all your updates, resolved all my individual comments and it all LGTM now, thanks for working on this! 🌟 

PS: This PR is now approved from my side, but please do check with @dmsnell before progressing with a merge. 🙏

dandominguez04 committed:
Abstracted toast method to remove duplicate code

dandominguez04 committed:
Added removeTag method in Note

Decouple the logic to remove a tag from a note from the repository and
move it to the model (Note).

dmsnell commented on Simplenote/src/main/java/com/automattic/simplenote/models/Note.java:
```diff
@@ -449,6 +449,14 @@ public void setPublished(boolean isPublished) {
 
+    public void removeTag(String tag) {
+        List<String> tags = getTags();
+        tags.remove(tag);
```
I'm sure you were planning on updating this as well, but we want to compare by tag hash here and not by tag name, or better, use some kind of `Tag.equals` call (which doesn't exist so probably best not to add in this PR)

```java
List<String> tags = getTags();
String tagHash = TagUtils.hashTag(tag);

for (String tagName : tags) {
    if (TagUtils.hashTag(tag) == tagHash) {
       tags.remove(tag);
    }
}

setTags(tags);
save();
```

dandominguez04 replied:
There are actually some methods in the `Note` model we can use to check tags:

```java
   public boolean hasTag(String tag) {
        List<String> tags = getTags();
        String tagLower = tag.toLowerCase();
        for (String tagName : tags) {
            if (tagLower.equals(tagName.toLowerCase())) return true;
        }
        return false;
    }

    public boolean hasTag(Tag tag) {
        return hasTag(tag.getSimperiumKey());
    }
```

Let me check how I can use them to check the tags.

dmsnell replied:
hm. I would recommend against using that method due to it being inaccurate - in separate work one of us should update that. also, it seems silly to iterate over all the tags inside a loop where we're already iterating over the tags.

I only meant to suggest that we rely on the `tagHash` here instead of the name. the whole `Tag.equals()` idea is just that basically. it wouldn't be more robust than…

```
// TagUtils.java or Tag.java
public static boolean areEqual(String left, String right) {
  return hashTag(left) == hashTag(right);
}
```

keeping it local and _not_ abstracting is probably most helpful at this point since `Tag`, `TagUtil`, and `Bucket` are all so jumbled and coupled together. in an ideal world I think that the `Note` object would have something like a `TagSet` property which itself held `Tag` objects and this would all be properly abstracted away from the data. for now it's just a jumble that would be hard to untangle 🤷‍♂️

dandominguez04 replied:
Yeah, I misread your first comment sorry (which had the solution). Fixed in a07747c09d8a98574723b9022e697c1f5a81f9e6.

dandominguez04 replied:
> in an ideal world I think that the Note object would have something like a TagSet property which itself held Tag objects and this would all be properly abstracted away from the data. for now it's just a jumble that would be hard to untangle

That would be interesting. It would be much simpler.

dandominguez04 committed:
Remove tags in notes comparing them with the canonical representation

The notes keep a list of tags associated. The tags saved are the name
of the tags and not their canonical representation. To remove the
correct tags, we need to compare the canonical representation of the
tag to remove against that canonical representation of the tags
associated with the notes.

dandominguez04 committed:
Remove characters from comments in TagUtils

danilo04 commented:
Hi @dmsnell 👋 . After talking with @ParaskP7, we decided to merge the PR to the feature branch and create separate issues for improvements on the tag management you mentioned.
